src/polyhedralGravity/model/SplitNode.h

```diff
@@ -0,0 +1,79 @@
+
+public:
+    SplitNode(const SplitParam &splitParam, Plane &plane, TriangleIndexLists &triangleIndexLists);
+    SplitNode(const SplitNode &other) = delete;
+    SplitNode(SplitNode &&other) = delete;
+    SplitNode &operator=(const SplitNode &other) = delete;
+    SplitNode &operator=(SplitNode &&other) = delete;
+    /**
+     * Computes the lesser child node if not present already and returns it to the caller.
+     * @return the SplitNode that is closer to the origin with respect to the split plane of this node.
+     */
+    TreeNode &getLesserNode();
+    /**
+     * Computes the greater child node if not present already and returns it to the caller.
+     * @return the SplitNode that is farther away of the origin with respect to the split plane of this node.
+     */
+    TreeNode &getGreaterNode();
+    /**
+     * Used to count the number of intersections of a ray and the polyhedron's faces.
+     * @param origin The point where the ray originates from.
+     * @param ray Specifies the ray direction.
+     * @return The number of intersections with the polyhedron.
+     */
+    unsigned long countIntersections(const polyhedralGravity::Array3 &origin, const polyhedralGravity::Array3 &ray) override;
+    void getFaceIntersections(const polyhedralGravity::Array3 &origin, const polyhedralGravity::Array3 &ray, std::vector<size_t> &intersectedFaceIndices) override;
+
+
+private:
+    /**
+     * Frees splitParam after both child nodes have been built.
+     */
+    void maybeFreeParam();
+    /**
+     * Calculates ray intersections with the bounding box using the concepts of slabs.
+     * @param origin The point where the ray originates from.
+     * @param ray Specifies the ray direction.
+     * @return Returns the t parameter for the entry and exit points of the ray with the box as a pair, with t being from the equation $intersection_point = orig + t * ray$.
+     */
+    [[nodiscard]] std::pair<double, double> rayBoxIntersection(const polyhedralGravity::Array3 &origin, const polyhedralGravity::Array3 &ray) const;
+    /**
+     * Intersects a ray with the splitPlane.
+     * @param origin The point where the ray originates from.
+     * @param ray Specifies the ray direction.
+     * @return Returns the t parameter for the intersection point, with t being from the equation $intersection_point = orig + t * ray$.
+     */
+    [[nodiscard]] double rayPlaneIntersection(const polyhedralGravity::Array3 &origin, const polyhedralGravity::Array3 &ray) const;
+    /**
+     * Gets the children of this node whose bounding boxes are hit by the ray.
+     * @param origin The point where the ray originates from.
+     * @param ray Specifies the ray direction.
+     * @return the child nodes that intersect with the ray
+     */
+    [[nodiscard]] std::unique_ptr<std::vector<TreeNode *>> getChildrenForIntersection(const polyhedralGravity::Array3 &origin, const polyhedralGravity::Array3 &ray) const;
```

Comment:
I see a legacy ghost 👻 here: `Tree Node*` --> Use a `shared_ptr` or `unique_ptr` or `std::optional` or a `const TreeNode&` whatever suits you better.